url270fc4f5ee_ykcilborw_Joust_tgz-pJ8-fri_patterns_interpreter_parsergenerator_examples_HowToOverrideLexerImplJ8/src/com/wroblicky/andrew/joust/game/chesspiece/Pawn.java

```diff
@@ -0,0 +1,184 @@
+package com.wroblicky.andrew.joust.game.chesspiece;
+
+import static com.wroblicky.andrew.joust.game.chesspiece.ChessPiece.Occupier.ENEMY;
+import static com.wroblicky.andrew.joust.game.chesspiece.ChessPiece.Occupier.UNOCCUPIED;
+
+import java.util.ArrayList;
+import java.util.List;
+
+import com.wroblicky.andrew.joust.game.board.ChessBoard;
+import com.wroblicky.andrew.joust.game.board.Location;
+import com.wroblicky.andrew.joust.game.subset.qualifiable.ChessPieceAllegianceType;
+import com.wroblicky.andrew.joust.game.subset.qualifiable.ChessPieceType;
+
+/**
+ * Represents the pawn chess piece
+ *
+ * @author Andrew Wroblicky
+ *
+ */
+public final class Pawn extends ChessPiece {
+
+	private boolean hasMovedAtAll;
+	private boolean movedTwice;
+
+	public Pawn() {
+		this.allegiance = Allegiance.BLACK;
+		this.hasMovedAtAll = false;
+		this.movedTwice = false;
+		this.alive = true;
+		this.chessID = "p1";
+		this.chessBoard = new ChessBoard();
+	}
+
+	public Pawn(Allegiance allegiance, int chessID, ChessBoard chessBoard) {
+		this.allegiance = allegiance;
+		this.hasMovedAtAll = false;
+		this.movedTwice = false;
+		this.alive = true;
+		this.chessID = getChessPieceAllegianceType()
+				.getChessPieceAllegianceType() + chessID;
+		this.chessBoard = chessBoard;
+	}
+
+	public boolean isMovedAtAll() {
+		return hasMovedAtAll;
+	}
+
+	public boolean isMovedTwice() {
+		return movedTwice;
+	}
+
+	@Override
+	public ChessPieceType getChessPieceType() {
+		return ChessPieceType.PAWN;
+	}
+
+	@Override
+	public ChessPieceAllegianceType getChessPieceAllegianceType() {
+		if (isBlack()) {
+			return ChessPieceAllegianceType.BLACK_PAWN;
+		}
+		return ChessPieceAllegianceType.WHITE_PAWN;
+	}
+
+	public void setMovedAtAll(boolean hasMovedAtAll) {
+		this.hasMovedAtAll = hasMovedAtAll;
+	}
+
+	public void setMovedTwice(boolean movedTwice) {
+		this.movedTwice = movedTwice;
+	}
+
+	// TODO
+	// add support for pawn promotion
+	@Override
+	public List<Location> getPossibleMoves() {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 17. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test.

We recommend that you simplify this method or break it into multiple methods. For example, consider extracting the code blocks on lines 95-105, lines 107-117 into separate methods.